fix memory leak

psychocoderHPC opened the pull request:
With #44 we tried to fix the memory leak which can happen if more than one thread is allocating a new chunk.
The PR #44 was not fixing the issue and introduced a double free issue.

- [ ] PR must be tested
    - [x] examples work on @psychocoderHPC  laptop  
- [x] this PR contains #47 else testing is not possible

@michaelsippel I have not tested the PR and wrote it blindly, we need to perform a runtime test before merging.

psychocoderHPC commented on redGrapes/util/chunked_list.hpp:
```diff
@@ -409,9 +406,14 @@ struct ChunkedList
     ChunkedList( size_t est_chunk_size = 32 )
         : ChunkedList(
             Allocator(),
-            est_chunk_size
```
@michaelsippel For me it is not clear why this constructor's implementation is different to the second constructor. Why do you not compute the `chunk_size` in the previous version before my PR.
This means for this constructor the member was initialized to zero.

michaelsippel replied:
The implementation is not different, the first one just uses the default allocator (picks the allocator assigned to the current worker) and is implemented by calling the second one

psychocoderHPC replied:
ohh yes, thanks for the clarification. I got confused by the words `chunk, chunks, Chunk, ...`^^ everywhere^^ (my fault)

michaelsippel commented on redGrapes/util/chunked_list.hpp:
```diff
@@ -163,14 +163,11 @@ struct ChunkedList
-         * compensated by not increasing the count when the last
-         * element is inserted.
+        /* counts the number of freed elements in this chunk.
+         * Whenever `freed_chunks` reaches the last chunk, the chunk
```
... reaches the last *item in this* chunk ...

michaelsippel commented on redGrapes/util/chunked_list.hpp:
```diff
@@ -163,14 +163,11 @@ struct ChunkedList
+         * will be deleted.
          */
-        std::atomic< chunk_offset_t > item_count{ 0 };
+        std::atomic< chunk_offset_t > freed_chunks{ 0 };
```
should be named accordingly  freed_items

psychocoderHPC replied:
renamed

michaelsippel commented on redGrapes/util/chunked_list.hpp:
```diff
@@ -453,25 +458,30 @@ struct ChunkedList
                     chunk->last_idx ++;
+                    if( chunk_off == chunk_size - 1)
+                    {
+                        // the thread who took the last chunk must allocate
```
...who took the last *item of this* chunk...

michaelsippel commented on redGrapes/util/chunked_list.hpp:
```diff
@@ -409,9 +406,14 @@ struct ChunkedList
+        size_t items_capacity = (chunks.get_chunk_capacity() - sizeof(Chunk));
+        this->chunk_size = items_capacity / sizeof(Item);
+        // start with one batch of chunks
+        chunks.allocate_item();
```
this ctor would allocate two chunks.

psychocoderHPC replied:
I reverted this

michaelsippel commented on redGrapes/util/chunked_list.hpp:
```diff
@@ -409,9 +406,14 @@ struct ChunkedList
         : ChunkedList(
             Allocator(),
-            est_chunk_size
+            est_chunk_size * sizeof(Item)
```
here we create chunks of size `est_chunk_size*sizeof(Item)*sizeof(Item)` .

This is actually what causes the CI error:
```
 Error: 2-11 16:40:48.450] [error] ChunkedBumpAlloc: requested allocation of 131072 bytes exceeds chunk capacity of 64992 byte
```

psychocoderHPC replied:
I removed it

michaelsippel commented on redGrapes/util/chunked_list.hpp:
```diff
@@ -453,25 +458,30 @@ struct ChunkedList
+                    if( chunk_off == chunk_size - 1)
+                    {
+                        // the thread who took the last chunk must allocate
+                        // the next batch of chunks
```
// THE next chunk
or
// the next batch of items

michaelsippel reviewed (commented):
generally there is a confusion between 'chunks' & 'items' in the variable naming & comments.  Otherwise I have a commit that refactors the Chunk such that it stores direct pointers instead of indices and we should build on that to implement the fix

psychocoderHPC committed:
fix broken example

readGrapes init and finalize is missing in `1_resource.cpp`

psychocoderHPC committed:
fix memory leak

With #44 we tried to fix the memory leak which can happen if more than one thread is allocating a new chunk.
The PR #44 was not fixing the issue and introduced a double free issue.

psychocoderHPC commented:
@michaelsippel I applied your review comments and rebased against #47

psychocoderHPC commented:
> generally there is a confusion between 'chunks' & 'items' in the variable naming & comments. Otherwise I have a commit that refactors the Chunk such that it stores direct pointers instead of indices and we should build on that to implement the fix

I am fine if we refactor these parts in a follow up but as long as there is no PR fix that instantly fixes the current issue of memory leaks I would like to bring it in to have a version that is working correctly.

I am also fine if this PR is not going in, both ways are possible as long as the dev branch becomes usable again.

michaelsippel commented:
For testing we should use small chunk sizes with many tasks since otherwise when everything fits into one chunk, the implementation will not be stressed

psychocoderHPC commented:
Unfortunately, this PR creates a deadlock :-( during testing

michaelsippel commented:
now already included in #48
